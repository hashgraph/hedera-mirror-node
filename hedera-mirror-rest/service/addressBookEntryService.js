/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2022 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

'use strict';

const _ = require('lodash');

const {AddressBook, AddressBookEntry, AddressBookServiceEndpoint} = require('../model');
const BaseService = require('./baseService');

/**
 * CryptoAllowance business model
 */
class AddressBookEntryService extends BaseService {
  static networkNodeseQuery = `with ${AddressBook.tableAlias} as (
    select ${AddressBook.START_CONSENSUS_TIMESTAMP}, ${AddressBook.FILE_ID}
    from ${AddressBook.tableName} where ${AddressBook.FILE_ID} = $1
   )
   select ${AddressBookEntry.getFullName(AddressBookEntry.DESCRIPTION)}, 
    ${AddressBook.getFullName(AddressBook.FILE_ID)}, 
    ${AddressBookEntry.getFullName(AddressBookEntry.MEMO)}, 
    ${AddressBookEntry.getFullName(AddressBookEntry.NODE_ID)}, 
    ${AddressBookEntry.getFullName(AddressBookEntry.NODE_ACCOUN_ID)}, 
    ${AddressBookEntry.getFullName(AddressBookEntry.NODE_CERT_HASH)},
    jsonb_agg(jsonb_build_object(
      '${AddressBookServiceEndpoint.IP_ADDRESS_V4}', ${AddressBookServiceEndpoint.IP_ADDRESS_V4}, 
      '${AddressBookServiceEndpoint.PORT}', ${AddressBookServiceEndpoint.PORT}
      ) 
      order by ${AddressBookServiceEndpoint.PORT} desc
    ) as service_endpoints, ${AddressBookEntry.getFullName(AddressBookEntry.PUBLIC_KEY)}
   from ${AddressBookEntry.tableName} ${AddressBookEntry.tableAlias}
   left join ${AddressBook.tableAlias} on 
    ${AddressBook.getFullName(AddressBook.START_CONSENSUS_TIMESTAMP)} = ${AddressBookEntry.getFullName(
    AddressBookEntry.CONSENSUS_TIMESTAMP
  )}
   left join ${AddressBookServiceEndpoint.tableName} ${AddressBookServiceEndpoint.tableAlias} on 
    ${AddressBook.getFullName(AddressBook.START_CONSENSUS_TIMESTAMP)} = ${AddressBookServiceEndpoint.getFullName(
    AddressBookServiceEndpoint.CONSENSUS_TIMESTAMP
  )}
   group by ${AddressBookEntry.getFullName(AddressBookEntry.DESCRIPTION)}, ${AddressBook.getFullName(
    AddressBook.FILE_ID
  )}, ${AddressBookEntry.getFullName(AddressBookEntry.MEMO)}, 
    ${AddressBookEntry.getFullName(AddressBookEntry.NODE_ID)}, ${AddressBookEntry.getFullName(
    AddressBookEntry.NODE_ACCOUN_ID
  )}, 
    ${AddressBookEntry.getFullName(AddressBookEntry.NODE_CERT_HASH)}, ${AddressBookEntry.getFullName(
    AddressBookEntry.PUBLIC_KEY
  )}`;

  async getNetworkNodes(conditions, initParams, order, limit) {
    const [query, params] = this.getNetworkNodesWithFiltersQuery(conditions, initParams, order, limit);

    const rows = await super.getRows(query, params, 'getNetworkNodes');
    return rows.map((ca) => new CryptoAllowance(ca));
  }

  getNetworkNodesWithFiltersQuery(whereConditions, whereParams, spenderOrder, limit) {
    const params = whereParams;
    const query = [
      AddressBookEntryService.networkNodeseQuery,
      whereConditions.length > 0 ? `where ${whereConditions.join(' and ')}` : '',
      super.getOrderByQuery(CryptoAllowance.SPENDER, spenderOrder),
      super.getLimitQuery(params.length + 1),
    ].join('\n');
    params.push(limit);

    return [query, params];
  }
}

module.exports = new AddressBookEntryService();
