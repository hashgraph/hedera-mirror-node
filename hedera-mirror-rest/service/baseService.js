/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import _ from 'lodash';

/**
 * Base service class that other services should inherit from for their retrieval business logic
 */
class BaseService {
  buildWhereSqlStatement(whereQuery, params = []) {
    if (_.isEmpty(whereQuery)) {
      return {where: '', params};
    }

    const prefix = params.length === 0 ? 'where' : 'and';
    const condition = whereQuery
      .map((current) => {
        let position;
        if (Array.isArray(current.param)) {
          const first = params.length + 1;
          const last = params.push(...current.param);
          position = `(${_.range(first, last + 1).map((pos) => '$' + pos)})`;
        } else {
          params.push(current.param);
          position = `$${params.length}`;
        }
        return `${current.query} ${position}`;
      })
      .join(' and ');

    return {where: `${prefix} ${condition}`, params};
  }

  getLimitQuery(position) {
    return `limit $${position}`;
  }

  /**
   * Gets the order by query from the order specs
   *
   * @param {OrderSpec} orderSpecs
   * @return {string}
   */
  getOrderByQuery(...orderSpecs) {
    return 'order by ' + orderSpecs.map((spec) => `${spec}`).join(', ');
  }

  getFilterWhereCondition(key, filter) {
    return {
      query: `${key} ${filter.operator}`,
      param: filter.value,
    };
  }

  async getRows(query, params, functionName = '') {
    if (logger.isTraceEnabled()) {
      logger.trace(`${functionName} query: ${query}, params: ${params}`);
    }
    const {rows} = await pool.queryQuietly(query, params);
    if (logger.isTraceEnabled()) {
      logger.trace(`${functionName} ${rows.length} entries`);
    }
    return rows;
  }

  async getSingleRow(query, params, functionName = '') {
    const rows = await this.getRows(query, params, functionName);
    if (_.isEmpty(rows) || rows.length > 1) {
      return null;
    }

    return rows[0];
  }

  /**
   * Builds a standard sql query that can be extended with an additional filters
   *
   * @param {string} selectFromTable
   * @param {*[]} params
   * @param {string[]} conditions
   * @param {string} orderClause
   * @param {string} limitClause
   * @param {{key: string, operator: string, value: *, column: string}[]} filters
   * @returns {(string|*)[]} the build query and the parameters for the query
   */
  buildSelectQuery(selectFromTable, params, conditions, orderClause, limitClause, filters = []) {
    const whereConditions = [
      ...conditions,
      ...filters.map((filter) => {
        params.push(filter.value);
        return `${filter.column}${filter.operator}$${params.length}`;
      }),
    ];

    return [
      selectFromTable,
      whereConditions.length > 0 ? `where ${whereConditions.join(' and ')}` : '',
      orderClause,
      limitClause,
    ].join('\n');
  }
}

export default BaseService;
