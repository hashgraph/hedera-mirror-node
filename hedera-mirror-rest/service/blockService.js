/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2022 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

'use strict';

const BaseService = require('./baseService');
const {RecordFile} = require('../model');

const buildWhereSqlStatement = (whereQuery) => {
  let where = '';
  const params = [];
  for (let i = 1; i <= whereQuery.length; i++) {
    where += `${i === 1 ? 'where' : 'and'} ${whereQuery[i - 1].query} $${i} `;
    params.push(whereQuery[i - 1].param);
  }

  return {where, params};
};

/**
 * BlockService retrieval business logic
 */
class BlockService extends BaseService {
  constructor() {
    super();
  }

  async getBlocks(filters) {
    const {where, params} = buildWhereSqlStatement(filters.whereQuery);

    const query = `
      select
        ${RecordFile.COUNT}, ${RecordFile.HASH}, ${RecordFile.NAME}, ${RecordFile.PREV_HASH}, ${RecordFile.BYTES},
        ${RecordFile.HAPI_VERSION_MAJOR}, ${RecordFile.HAPI_VERSION_MINOR}, ${RecordFile.HAPI_VERSION_PATCH},
        ${RecordFile.INDEX}, ${RecordFile.CONSENSUS_START}, ${RecordFile.CONSENSUS_END}
      from ${RecordFile.tableName}
      ${where}
      order by ${RecordFile.INDEX} ${filters.order}
      limit ${filters.limit}
    `;
    const rows = await super.getRows(query, params, 'getBlocks');

    return rows.map((recordFile) => new RecordFile(recordFile));
  }
}

module.exports = new BlockService();
