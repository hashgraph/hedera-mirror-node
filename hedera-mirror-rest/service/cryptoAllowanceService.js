/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import BaseService from './baseService';
import {CryptoAllowance} from '../model';
import {OrderSpec} from '../sql';

/**
 * CryptoAllowance business model
 */
class CryptoAllowanceService extends BaseService {
  static accountAllowanceQuery = `select * from ${CryptoAllowance.tableName}`;

  async getAccountCryptoAllowances(conditions, initParams, order, limit) {
    const {query, params} = this.getAccountAllowancesQuery(conditions, initParams, order, limit);
    const rows = await super.getRows(query, params, 'getAccountCryptoAllowances');
    return rows.map((ca) => new CryptoAllowance(ca));
  }

  getAccountAllowancesQuery(whereConditions, whereParams, order, limit) {
    const params = whereParams;
    params.push(limit);
    const query = [
      CryptoAllowanceService.accountAllowanceQuery,
      whereConditions.length > 0 ? `where ${whereConditions.join(' and ')}` : '',
      super.getOrderByQuery(OrderSpec.from(CryptoAllowance.SPENDER, order)),
      super.getLimitQuery(params.length),
    ].join('\n');

    return {query, params};
  }
}

export default new CryptoAllowanceService();
