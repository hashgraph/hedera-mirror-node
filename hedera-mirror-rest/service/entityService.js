/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import _ from 'lodash';

import AccountAlias from '../accountAlias';
import BaseService from './baseService';
import {filterKeys} from '../constants';
import EntityId from '../entityId';
import {InvalidArgumentError, NotFoundError} from '../errors';
import {Entity} from '../model';

/**
 * Entity retrieval business logic
 */
class EntityService extends BaseService {
  static entityFromAliasQuery = `select ${Entity.ID}
                                 from ${Entity.tableName}`;

  static entityFromEvmAddressQuery = `select ${Entity.ID}
                                      from ${Entity.tableName}
                                      where ${Entity.DELETED} <> true
                                        and ${Entity.EVM_ADDRESS} = $1`;

  // use a small column in existence check to reduce return payload size
  static entityExistenceQuery = `select ${Entity.TYPE}
                                 from ${Entity.tableName}
                                 where ${Entity.ID} = $1`;

  static aliasColumns = [Entity.SHARD, Entity.REALM, Entity.ALIAS];
  static aliasConditions = [`coalesce(${Entity.DELETED}, false) <> true`];
  static missingAccountAlias = 'No account with a matching alias found';
  static multipleAliasMatch = `Multiple alive entities matching alias`;
  static multipleEvmAddressMatch = `Multiple alive entities matching evm address`;

  constructor() {
    super();
  }

  /**
   * Retrieves the entity containing matching the given alias
   *
   * @param {AccountAlias} accountAlias accountAlias
   * @return {Promise<Entity>} raw entity object
   */
  async getAccountFromAlias(accountAlias) {
    const params = [];
    const conditions = [].concat(EntityService.aliasConditions);

    EntityService.aliasColumns
      .filter((column) => accountAlias[column] !== null)
      .forEach((column) => {
        const length = params.push(accountAlias[column]);
        conditions.push(`${column} = $${length}`);
      });

    const aliasQuery = `${EntityService.entityFromAliasQuery} where ${conditions.join(' and ')}`;

    const rows = await super.getRows(aliasQuery, params, 'getAccountFromAlias');

    if (_.isEmpty(rows)) {
      return null;
    } else if (rows.length > 1) {
      logger.error(`Incorrect db state: ${rows.length} alive entities matching alias ${accountAlias}`);
      throw new Error(EntityService.multipleAliasMatch);
    }

    return new Entity(rows[0]);
  }

  /**
   * Checks if provided accountId maps to a valid entity
   * @param {String} accountId
   * @returns {Promise} valid flag
   */
  async isValidAccount(accountId) {
    const entity = await super.getSingleRow(EntityService.entityExistenceQuery, [accountId], 'isValidAccount');
    return !_.isNil(entity);
  }

  /**
   * Gets the encoded account id from the account alias string.
   * @param {AccountAlias} accountAlias the account alias object
   * @return {Promise<BigInt>}
   */
  async getAccountIdFromAlias(accountAlias) {
    const entity = await this.getAccountFromAlias(accountAlias);
    if (_.isNil(entity)) {
      throw new NotFoundError(EntityService.missingAccountAlias);
    }

    return entity.id;
  }

  /**
   * Gets the encoded entity id from the evm address.
   *
   * @param {String} evmAddress
   * @return {Promise<BigInt|Number>}
   */
  async getEntityIdFromEvmAddress(evmAddress) {
    const rows = await this.getRows(
      EntityService.entityFromEvmAddressQuery,
      Buffer.from(evmAddress, 'hex'),
      'getEntityIdFromEvmAddress'
    );
    if (rows.length === 0) {
      throw new NotFoundError();
    } else if (rows.length > 1) {
      logger.error(`Incorrect db state: ${rows.length} alive entities matching evm address ${evmAddress}`);
      throw new Error(EntityService.multipleEvmAddressMatch);
    }

    return rows[0].id;
  }

  /**
   * Retrieve the encodedId of a validated EntityId from a shard.realm.num string, encoded id string, evm address with
   * optional shard and realm, or alias string.
   * Throws {@link InvalidArgumentError} if the entity id string is invalid
   * Throws {@link NotFoundError} if the account is not present when retrieving by alias or the entity is not present
   * when retrieving by evm address
   *
   * @param {String} entityIdString
   * @param {String} paramName the parameter name
   * @returns {Promise} entityId
   */
  async getEncodedId(entityIdString, paramName = filterKeys.ID_OR_ALIAS_OR_EVM_ADDRESS) {
    try {
      if (EntityId.isValidEntityId(entityIdString)) {
        const entityId = EntityId.parse(entityIdString, {paramName});
        return entityId.evmAddress === null
          ? entityId.getEncodedId()
          : await this.getEntityIdFromEvmAddress(entityId.evmAddress);
      } else if (AccountAlias.isValid(entityIdString)) {
        return await this.getAccountIdFromAlias(AccountAlias.fromString(entityIdString));
      }
    } catch (ex) {
      if (ex instanceof InvalidArgumentError) {
        throw InvalidArgumentError.forParams(paramName);
      }
      // rethrow
      throw ex;
    }

    throw InvalidArgumentError.forParams(paramName);
  }
}

export default new EntityService();
