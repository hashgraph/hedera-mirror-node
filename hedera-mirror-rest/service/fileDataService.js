/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import _ from 'lodash';

import BaseService from './baseService';
import {ExchangeRate, FileData, FeeSchedule} from '../model';

/**
 * File data retrieval business logic
 */
class FileDataService extends BaseService {
  static exchangeRateFileId = 112;
  static feeScheduleFileId = 111;

  // placeholders to support where filtering for inner and outer calls
  static filterInnerPlaceholder = '<filterInnerPlaceHolder>';
  static filterOuterPlaceholder = '<filterOuterPlaceHolder>';

  // retrieve the largest timestamp of the most recent create/update operation on the file
  // using this timestamp retrieve all recent file operations and combine contents for applicable file
  static latestFileContentsQuery = `with latest_create as (
      select max(${FileData.CONSENSUS_TIMESTAMP}) as ${FileData.CONSENSUS_TIMESTAMP}
      from ${FileData.tableName}
      where ${FileData.ENTITY_ID} = $1 and ${FileData.TRANSACTION_TYPE} in (17, 19) ${
    FileDataService.filterInnerPlaceholder
  }
      group by ${FileData.ENTITY_ID}
      order by ${FileData.CONSENSUS_TIMESTAMP} desc
    )
    select
      max(${FileData.tableAlias}.${FileData.CONSENSUS_TIMESTAMP}) as ${FileData.CONSENSUS_TIMESTAMP},
      string_agg(${FileData.getFullName(FileData.FILE_DATA)}, '' order by ${FileData.getFullName(
    FileData.CONSENSUS_TIMESTAMP
  )}) as ${FileData.FILE_DATA}
    from ${FileData.tableName} ${FileData.tableAlias}
    join latest_create l on ${FileData.getFullName(FileData.CONSENSUS_TIMESTAMP)} >= l.${FileData.CONSENSUS_TIMESTAMP}
    where ${FileData.getFullName(FileData.ENTITY_ID)} = $1 and ${FileData.getFullName(
    FileData.TRANSACTION_TYPE
  )} in (16,17, 19)
      and ${FileData.getFullName(FileData.CONSENSUS_TIMESTAMP)} >= l.${FileData.CONSENSUS_TIMESTAMP} ${
    FileDataService.filterOuterPlaceholder
  }
    group by ${FileData.getFullName(FileData.ENTITY_ID)}`;

  static getFileDataQuery = `select
         string_agg(
           ${FileData.FILE_DATA}, ''
           order by ${FileData.CONSENSUS_TIMESTAMP}
           ) data
        from ${FileData.tableName}
        where
           ${FileData.ENTITY_ID} = $1
        and ${FileData.CONSENSUS_TIMESTAMP} >= (
        select ${FileData.CONSENSUS_TIMESTAMP}
        from ${FileData.tableName}
        where ${FileData.ENTITY_ID} = $1
        and ${FileData.CONSENSUS_TIMESTAMP} <= $2
        and (${FileData.TRANSACTION_TYPE} = 17
             or ( ${FileData.TRANSACTION_TYPE} = 19
                  and
                  length(${FileData.FILE_DATA}) <> 0 ))
        order by ${FileData.CONSENSUS_TIMESTAMP} desc
        limit 1
        ) and ${FileData.CONSENSUS_TIMESTAMP} <= $2`;

  /**
   * The function returns the data for the fileId at the provided consensus timestamp.
   * @param fileId
   * @param timestamp
   * @return {data: string}
   */
  getFileData = async (fileId, timestamp) => {
    const params = [fileId, timestamp];
    const query = FileDataService.getFileDataQuery;
    const row = await super.getSingleRow(query, params, 'getFileData');
    return _.isNil(row) ? null : row.data;
  };

  getLatestFileContentsQuery = (innerWhere = '') => {
    const outerWhere = innerWhere.replace('and ', `and ${FileData.tableAlias}.`);
    return FileDataService.latestFileContentsQuery
      .replace(FileDataService.filterInnerPlaceholder, innerWhere)
      .replace(FileDataService.filterOuterPlaceholder, outerWhere);
  };

  getLatestFileDataContents = async (fileId, filterQueries) => {
    const {where, params} = super.buildWhereSqlStatement(filterQueries.whereQuery, [fileId]);
    return super.getSingleRow(this.getLatestFileContentsQuery(where), params, 'getLatestFileContents');
  };

  getExchangeRate = async (filterQueries) => {
    const row = await this.getLatestFileDataContents(FileDataService.exchangeRateFileId, filterQueries);
    return _.isNil(row) ? null : new ExchangeRate(row);
  };

  getFeeSchedule = async (filterQueries) => {
    const row = await this.getLatestFileDataContents(FileDataService.feeScheduleFileId, filterQueries);
    return _.isNil(row) ? null : new FeeSchedule(row);
  };
}

export default new FileDataService();
