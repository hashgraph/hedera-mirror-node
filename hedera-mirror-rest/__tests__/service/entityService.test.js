/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import {EntityService} from '../../service';
import AccountAlias from '../../accountAlias';
import integrationDomainOps from '../integrationDomainOps';
import {setupIntegrationTest} from '../integrationUtils';

setupIntegrationTest();

const defaultEntityAlias = new AccountAlias('1', '2', 'KGNABD5L3ZGSRVUCSPDR7TONZSRY3D5OMEBKQMVTD2AC6JL72HMQ');
const defaultInputEntity = [
  {
    alias: defaultEntityAlias.base32Alias,
    evm_address: 'ac384c53f03855fa1b3616052f8ba32c6c2a2fec',
    id: 281483566645248,
    num: 0,
    shard: 1,
    realm: 2,
  },
];
const defaultInputContract = [
  {
    evm_address: 'cef2a2c6c23ab8f2506163b1af55830f35c483ca',
    id: 4295062919,
    num: 95623,
    shard: 0,
    realm: 1,
  },
];

const defaultExpectedEntity = {id: 281483566645248};
const defaultExpectedContractId = {id: 4295062919};

describe('EntityService.getAccountFromAlias tests', () => {
  test('EntityService.getAccountFromAlias - No match', async () => {
    await expect(EntityService.getAccountFromAlias({alias: '1'})).resolves.toBeNull();
  });

  test('EntityService.getAccountFromAlias - Matching entity', async () => {
    await integrationDomainOps.loadEntities(defaultInputEntity);

    await expect(EntityService.getAccountFromAlias(defaultEntityAlias)).resolves.toMatchObject(defaultExpectedEntity);
  });

  test('EntityService.getAccountFromAlias - Duplicate alias', async () => {
    const inputEntities = [
      {
        alias: defaultEntityAlias.base32Alias,
        id: 3,
        num: 3,
        shard: 1,
        realm: 2,
      },
      {
        alias: defaultEntityAlias.base32Alias,
        id: 4,
        num: 4,
        shard: 1,
        realm: 2,
      },
    ];
    await integrationDomainOps.loadEntities(inputEntities);

    await expect(() => EntityService.getAccountFromAlias(defaultEntityAlias)).rejects.toThrowErrorMatchingSnapshot();
  });
});

describe('EntityService.getAccountIdFromAlias tests', () => {
  test('EntityService.getAccountIdFromAlias - No match', async () => {
    await expect(() => EntityService.getAccountIdFromAlias(defaultEntityAlias)).rejects.toThrowErrorMatchingSnapshot();
  });

  test('EntityService.getAccountFromAlias - Matching id', async () => {
    await integrationDomainOps.loadEntities(defaultInputEntity);

    await expect(EntityService.getAccountIdFromAlias(defaultEntityAlias)).resolves.toBe(defaultExpectedEntity.id);
  });
});

describe('EntityService.getEntityIdFromEvmAddress tests', () => {
  const defaultEvmAddress = defaultInputEntity[0].evm_address;

  test('EntityService.getEntityIdFromEvmAddress - Matching evm address', async () => {
    await integrationDomainOps.loadEntities(defaultInputEntity);

    await expect(EntityService.getEntityIdFromEvmAddress(defaultEvmAddress)).resolves.toBe(defaultExpectedEntity.id);
  });

  test('EntityService.getEntityIdFromEvmAddress - No match', async () => {
    await expect(() =>
      EntityService.getEntityIdFromEvmAddress(defaultEvmAddress)
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  test('EntityService.getEntityIdFromEvmAddress - Multiple matches', async () => {
    const inputEntities = [
      defaultInputEntity[0],
      {
        ...defaultInputEntity[0],
        id: defaultInputEntity[0].id + 1,
        num: defaultInputEntity[0].num + 1,
      },
    ];
    await integrationDomainOps.loadEntities(inputEntities);

    await expect(() =>
      EntityService.getEntityIdFromEvmAddress(defaultEvmAddress)
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});

describe('EntityService.isValidAccount tests', () => {
  test('EntityService.isValidAccount - No match', async () => {
    await expect(EntityService.isValidAccount(defaultInputEntity[0].id)).resolves.toBe(false);
  });

  test('EntityService.getAccountFromAlias - Matching', async () => {
    await integrationDomainOps.loadEntities(defaultInputEntity);

    await expect(EntityService.isValidAccount(defaultInputEntity[0].id)).resolves.toBe(true);
  });
});

describe('EntityService.getEncodedId tests', () => {
  test('EntityService.getEncodedId - No match', async () => {
    await expect(EntityService.getEncodedId(defaultInputEntity[0].id)).resolves.toBe(defaultExpectedEntity.id);
  });

  test('EntityService.getEncodedId - Matching id', async () => {
    await expect(EntityService.getEncodedId(defaultInputEntity[0].id)).resolves.toBe(defaultExpectedEntity.id);
    await expect(EntityService.getEncodedId(defaultInputContract[0].id)).resolves.toBe(defaultExpectedContractId.id);
  });

  test('EntityService.getEncodedId - Matching alias', async () => {
    await integrationDomainOps.loadEntities(defaultInputEntity);

    await expect(EntityService.getEncodedId(defaultInputEntity[0].alias)).resolves.toBe(defaultExpectedEntity.id);
  });

  test('EntityService.getEncodedId - Matching evm address', async () => {
    await integrationDomainOps.loadEntities(defaultInputEntity);
    await integrationDomainOps.loadContracts(defaultInputContract);

    const accountEvmAddress = defaultInputEntity[0].evm_address;
    await expect(EntityService.getEncodedId(accountEvmAddress)).resolves.toBe(defaultExpectedEntity.id);
    await expect(EntityService.getEncodedId(`0x${accountEvmAddress}`)).resolves.toBe(defaultExpectedEntity.id);

    const contractEvmAddress = defaultInputContract[0].evm_address;
    await expect(EntityService.getEncodedId(contractEvmAddress)).resolves.toBe(defaultExpectedContractId.id);
    await expect(EntityService.getEncodedId(`0x${contractEvmAddress}`)).resolves.toBe(defaultExpectedContractId.id);
  });

  test('EntityService.getEncodedId - Invalid alias', async () => {
    await expect(EntityService.getEncodedId('deadbeef=')).rejects.toThrowErrorMatchingSnapshot();
  });
});
