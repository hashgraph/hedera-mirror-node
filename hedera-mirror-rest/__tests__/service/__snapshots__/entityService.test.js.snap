// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityService.getAccountFromAlias tests EntityService.getAccountFromAlias - Duplicate alias 1`] = `"Multiple alive entities matching alias"`;

exports[`EntityService.getAccountIdFromAlias tests EntityService.getAccountIdFromAlias - No match 1`] = `"No account with a matching alias found"`;

exports[`EntityService.getEncodedId tests EntityService.getEncodedId - Invalid alias 1`] = `
[
  "Invalid parameter: idOrAliasOrEvmAddress",
]
`;

exports[`EntityService.getEntityIdFromEvmAddress tests EntityService.getEntityIdFromEvmAddress - Multiple matches 1`] = `"Multiple alive entities matching evm address"`;

exports[`EntityService.getEntityIdFromEvmAddress tests EntityService.getEntityIdFromEvmAddress - No match 1`] = `"Not found"`;
